utils/slot.mjs

```diff
@@ -1,45 +1,29 @@
-import { sleep } from "./misc.mjs";
 import { rpcSubscriptions } from "./rpc.mjs";
 
 const MAX_SLOT_FETCH_ATTEMPTS = process.env.MAX_SLOT_FETCH_ATTEMPTS || 100;
-let attempts = 0;
-const abortController = new AbortController();
 
-export const watchSlotSent = async (gSlotSent) => {
-  try {
-    const slotNotifications = await rpcSubscriptions
-      .slotsUpdatesNotifications()
-      .subscribe({ abortSignal: abortController.signal });
-
-    for await (const notification of slotNotifications) {
-      if (notification.type === "firstShredReceived") {
-        gSlotSent.value = notification.slot;
-        gSlotSent.updated_at = Date.now();
-        attempts = 0;
-        continue;
-      }
-
-      gSlotSent.value = null;
-      gSlotSent.updated_at = 0;
-
-      ++attempts;
-
-      if (attempts >= MAX_SLOT_FETCH_ATTEMPTS) {
-        console.log(
-          `ERROR: Max attempts for fetching slot type "firstShredReceived" reached, exiting`
-        );
-        process.exit(0);
-      }
-
-      // If update not received in last 3s, re-subscribe
-      if (gSlotSent.value !== null) {
-        while (Date.now() - gSlotSent.updated_at < 3000) {
-          await sleep(1);
-        }
-      }
+let slotNotifications;
+async function* slots() {
+  slotNotifications ||= await rpcSubscriptions
+    .slotsUpdatesNotifications()
+    .subscribe({ abortSignal: AbortSignal.any([]) });
+  let attempts = 0;
+  for await (const notification of slotNotifications) {
+    if (notification.type === "firstShredReceived") {
+      attempts = 0;
+      yield notification.slot;
+      continue;
+    }
+    if (++attempts >= MAX_SLOT_FETCH_ATTEMPTS) {
+      console.log(
+        `ERROR: Max attempts for fetching slot type "firstShredReceived" reached, exiting`
+      );
+      process.exit(0);
     }
-  } catch (e) {
-    console.log(`ERROR: ${e}`);
-    ++attempts;
   }
-};
+}
+
+export async function getNextSlot() {
+  const { value: nextSlot } = await slots().next()
+  return nextSlot;
+}
```

Comment:
This really does give you the _next_ slot when you call it, not the current one, because I haven't set this up as a read-through cache. I think this is consistent with your original implementation.